Improve getSpecificPrice static cache for products with combinations

Sollsy opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | getSpecificPrice is called on every combination for product and caching one result at a time. Instead we should call it once for all product variants and cache them. This way we reduce memory usage, query doubles and loading speed. Specific prices is allways returned with results from query or false if no match found. The more combinations product has the bigger improvement you will notice.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | create product with a lot of combinations 50-100+. Below you can find improvements on product with 700 combinations. 
| UI Tests          |  CI test are green
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | 

Tested on 9.0.x,8.2.x,1.7.8.x.  Below you can see difference in loading product page with ~700 combinations, which looks like that from customer point of view.
<img width="532" height="428" alt="image" src="https://github.com/user-attachments/assets/84463644-5a88-4d9e-b1aa-7a80bb6512dc" />

Before change:
<img width="1679" height="250" alt="image" src="https://github.com/user-attachments/assets/0a00aaa3-885c-427b-a802-60663d6a0280" />

After change:
<img width="1664" height="255" alt="image" src="https://github.com/user-attachments/assets/46910354-1265-4573-92a7-4df8d2b79848" />

Sollsy committed:
Update SpecificPrice.php

Improve getSpecificPrice static cache for products with combinations

Sollsy committed:
Update SpecificPrice.php

csfixer

Sollsy committed:
Update SpecificPrice.php

csfixer

Sollsy committed:
Update SpecificPrice.php

csfixer

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @Sollsy!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

Hlavtox reviewed (changes requested):
This can hurt performance in many contexts.

You are calling product::getProductAttributesIds just to prefill a cache, even if the caller is possibly interested in one price only.

But yeah, it could definitely be the way to fetch all specific in one query right away. It would possibly also make easier determining the order of specific prices, without needing to make more queries. :-)

Sollsy committed:
Update SpecificPrice.php

allow caller to decide if he wants to get single or all specificprices for the product

Sollsy commented:
> This can hurt performance in many contexts.
> 
> You are calling product::getProductAttributesIds just to prefill a cache, even if the caller is possibly interested in one price only.
> 
> But yeah, it could definitely be the way to fetch all specific in one query right away. It would possibly also make easier determining the order of specific prices, without needing to make more queries. :-)

It all depends on the amount of products and specific prices applied. I've been using this for over 3 months and it did speedup loading with a catalog of 40k product variants (7-10k active products atleast half with multiple specific prices).

If store has up to 10 combinations per product it wont be huge improvement, but once you hit 100-200 it does matter. Anyway i added variable so the caller could decide if he wants only single value or all.

Sollsy committed:
Update SpecificPrice.php

Sollsy committed:
Update SpecificPrice.php

Sollsy committed:
Update SpecificPrice.php

Sollsy committed:
Update Product.php
